File: Sequence/q1-q99/q30-q39/q39/q39.py
class Solution:
    def dfs(self, path, solution, target, index, candidates):
        if target == 0:
            solution.append(path[:])
            return
        for i in range(index, len(candidates)):
            if candidates[i] > target:
                return
            path.append(candidates[i])
            self.dfs(path, solution, target-candidates[i], i,candidates)
            path.pop()
            
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        path = []
        solution = []
        candidates.sort()
        self.dfs(path, solution, target, 0, candidates)
        return solution

File: Sequence/q1-q99/q30-q39/q39/q39.cpp
#include <algorithm>
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        std::vector<int> path;
        std::vector<std::vector<int>> solution;
        std::sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0, path, solution);
        return solution;
    }
private:
    void dfs(vector<int>& candidates, int target, int index, vector<int>& path, vector<vector<int>>& solution){
        if (target == 0){
            solution.push_back(path);
            return;
        }
        for (int i = index; i < candidates.size(); i++){
            if (candidates[i] > target)
                return;
            path.push_back(candidates[i]);
            dfs(candidates, target-candidates[i], i, path, solution);
            path.pop_back();
        }
    }
};
